In [34]:
def AVG(lst):
    new_lst=[]
    for i in range(10,len(lst)-11):
        new_lst.append(sum(lst[i-10:i+10])/20)
    return new_lst

In [265]:
import pandas 
import numpy as np
import os
import matplotlib
import IPython
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import axes3d  
from mpl_toolkits.axes_grid1.inset_locator import zoomed_inset_axes 
from mpl_toolkits.axes_grid1.inset_locator import mark_inset
import time
path = r"C:\Users\forma\Programming\Jupyter Projects\מעבדה ב'\זהירה תרמית"
dir_list = os.listdir()
os.chdir(path)



def PRINTDATA(n):
    TEMP=[]
    VOLT=[]
    LIG=[]
    count=0
    try:
            with open('{}sec.dat'.format(n)) as f:
                for line in f.readlines():
                    line=line.split()
                    if count == 0:
                        count=+1
                        continue
                    else:
                        
                        CUTBEG=140
                        CUTEND=-60
                        TEMP.append(float(line[0]))
                        VOLT.append(float(line[1]))
                        LIG.append(float(i))
                        
            TEMPAVG = np.array(AVG(TEMP[CUTBEG:CUTEND]))
            
            VOLTAVG = np.array(AVG(VOLT[CUTBEG:CUTEND]))
            
            #TEMP = np.array(TEMP[CUTBEG:CUTEND])
            #VOLT = np.array(VOLT[CUTBEG:CUTEND])
            #LIG = np.array(LIG[0:len(Volt)+1])
            #print(TEMPAVG)
            #print(VOLTAVG)
            return TEMPAVG,VOLTAVG
        
            #fig, ax1 = plt.subplots(figsize=(13,5))
            #plt.plot(range(10), linestyle='--', marker='.', color='b')
            #color = 'tab:red'
            #ax1.set(title='Linear Region "Cutting"'.format(n))
            #ax1.set_xlabel(r'$TEMP$')
            #ax1.set_ylabel('Data[V]', color=color)
            #ax1.plot(TEMPAVG, VOLTAVG, color=color)
            #ax1.tick_params(axis='y', labelcolor=color)
            # draw vertical line from (70,100) to (70, 250)
            #plt.plot([425, 490], [0.17, 0.17], 'k-', lw=2)
            #plt.plot([410, 500], [0.09, 0.09], 'k-', lw=2)
            #plt.text(431.5, 0.12, r'Supralinear Region', fontsize=20)
            #I want to select the x-range for the zoomed region. I have figured it out suitable values
            # by trial and error. How can I pass more elegantly the dates as something like
            x1 = 400
            x2 = 411
            
            # select y-range for zoomed region
            y1 = 0.02
            y2 = 0.05
            
            # Make the zoom-in plot:
            #axins = zoomed_inset_axes(ax1, 2, loc=1) # zoom = 2
            #axins.plot(TEMP, VOLT, color=color)
            #axins.set_xlim(x1, x2)
            #axins.set_ylim(y1, y2)
            #plt.xticks(visible=False)
            #plt.yticks(visible=False)
            #mark_inset(ax1, axins, loc1=2, loc2=4, fc="none", ec="0.5")
            #plt.draw()
            #ax2 = ax1.twinx()  # instantiate a second axes that shares the same x-axis
            #color = 'tab:blue'
            #ax2.set_ylabel(r'$VOLT AVG$', color=color)  # we already handled the x-label with ax1
            #ax2.plot(TEMPAVG, VOLTAVG, color=color)
            #ax2.tick_params(axis='y', labelcolor=color)
            
            #fig.tight_layout()  # otherwise the right y-label is slightly clipped
            #fig.savefig(r"G:\My Drive\לימודים\תשפ''א ב\פיזיקה תרמית\שיעורי בית\ERRORS.png")
            #plt.show()
            #ax.plot(TEMP, VOLT)
    
            #ax.set(title='Voltage(T),Lightning time:{}sec'.format(n))
           
            #fig.savefig(r"C:\Users\forma\Programming\Jupyter Projects\מעבדה ב'\זהירה תרמית\MEASURMENTS\{}secREGIONSCUTT.png".format(n))
            #count=0
            #plt.show()
            #f.close()
    except:
        return None 
#for i in range(60,58):
    #PRINTDATA(i)
TEMP2=[]
VOLT2=[]
LIG2=[]
for i in range(1,500):
    if PRINTDATA(i) == None:

        continue
    else:
        T,V=PRINTDATA(i)
        TEMP2=T
        VOLT2=V
        #LIG2=L

        TEMP2 = np.array(TEMP2)
        VOLT2 = np.array(VOLT2)
        #print(TEMP2)
        #print(VOLT2)
        #LIG2 = np.array(LIG2)
        print('For EXP time:{}sec'.format(i))
        FHALF(VOLT2,TEMP2)



For EXP time:30sec
$T_1$ is: 422.26054999999997 
VOLT in $T_1$ is: 0.04193690000000001 
$T_2$ is: 486.7296499999999 
VOLT in $T_2$ is: 0.04274075 
$T_m$ is: 458.1823499999999 
VOLT $T_m$ is: 0.08510614999999999 
$T_2-T_1$=64.46909999999991 
$T_2-T_m$=28.547300000000007 
$(T_2-T_m)/(T_2-T_1)$=0.4428059333851418 

For EXP time:54sec
$T_1$ is: 426.36455000000007 
VOLT in $T_1$ is: 0.11353689999999998 
$T_2$ is: 482.9153000000001 
VOLT in $T_2$ is: 0.11709675000000001 
$T_m$ is: 457.98525000000006 
VOLT $T_m$ is: 0.23162735000000004 
$T_2-T_1$=56.55075000000005 
$T_2-T_m$=24.93005000000005 
$(T_2-T_m)/(T_2-T_1)$=0.44084384380401725 

For EXP time:60sec
$T_1$ is: 426.28799999999984 
VOLT in $T_1$ is: 0.09245025 
$T_2$ is: 482.43494999999996 
VOLT in $T_2$ is: 0.09609279999999998 
$T_m$ is: 458.12885000000006 
VOLT $T_m$ is: 0.18578665 
$T_2-T_1$=56.14695000000012 
$T_2-T_m$=24.3060999999999 
$(T_2-T_m)/(T_2-T_1)$=0.4329015200291352 

For EXP time:90sec
$T_1$ is: 426.96504999999996 
VOLT in 

In [264]:
def FHALF(array2,array1):
    MAXMAX=max(array2)
    indexmaxvolt=np.where(array2==max(array2))
    HALFMAX=MAXMAX/2
    
    for i in range(len(array2)-1):
        
        if HALFMAX > array2[i] and  HALFMAX < array2[i+1]:
            FIRHALF = i

        if HALFMAX < array2[i] and  HALFMAX > array2[i+1]:
            SECHALF = i
        else:
            continue

    print('$T_1$ is: {}'.format(TEMP2[FIRHALF]),
          '\nVOLT in $T_1$ is: {}'.format(VOLT2[FIRHALF]),
          '\n$T_2$ is: {}'.format(TEMP2[SECHALF]),
          '\nVOLT in $T_2$ is: {}'.format(VOLT2[SECHALF]),
          '\n$T_m$ is: {}'.format(float(TEMP2[indexmaxvolt])),
         '\nVOLT $T_m$ is: {}'.format(MAXMAX),
          '\n$T_2-T_1$={}'.format(TEMP2[SECHALF]-TEMP2[FIRHALF]),
          '\n$T_2-T_m$={}'.format(TEMP2[SECHALF]-float(TEMP2[indexmaxvolt])),
          '\n$(T_2-T_m)/(T_2-T_1)$={}'.format((TEMP2[SECHALF]-float(TEMP2[indexmaxvolt]))/(TEMP2[SECHALF]-TEMP2[FIRHALF])),'\n')

In [157]:
def MAXPOINT(array_y,array_x):
    mxindex=np.argmax(array_y)
    return array_x[mxindex]

In [158]:
max_points=[]
for i in range(30,500):
    if PRINTDATA(i) == None:

        continue
    else:
        T,V=PRINTDATA(i)
        max_points.append(MAXPOINT(V,T))
max_points=np.array(max_points)
print("Avrage maxima is:",sum(max_points)/len(max_points),
      '\nand the biggest spread from avg is {} and {}'.format(min(max_points),max(max_points)))

Avrage maxima is: 458.02381818181817 
and the biggest spread from avg is 457.0970000000001 and 459.15795000000014


In [160]:
def FHALF(array1,array2):
    HALFMAX=max(array2)/2
    for i in range(len(array2)-1):
        if HALFMAX > array2[i] and  HALFMAX < array2[i+1]:
            FIRHALF = i
        if HALFMAX < array2[i] and  HALFMAX > array2[i+1]:
            SECHALF = i
        
        else:
            continue

    print('Temprature in Half max first point is: {}'.format(TEMP2[FIRHALF]),
          '\nVOLT in Half max first point is: {}'.format(VOLT2[SECHALF]),
          '\nTemprature in Half max second point is: {}'.format(TEMP2[FIRHALF]),
          '\nVOLT in Half max second point is: {}'.format(VOLT2[SECHALF]))
  
 

For EXP time: 30sec    
$T_1$ is: 422.26054999999997   
VOLT in $T_1$ is: 0.04193690000000001   
$T_2$ is: 486.7296499999999   
VOLT in $T_2$ is: 0.04274075   
$T_m$ is: 458.1823499999999   
VOLT $T_m$ is: 0.08510614999999999   
$T_2-T_1$=64.46909999999991   
$T_2-T_m$=28.547300000000007   
$(T_2-T_m)/(T_2-T_1)$=0.4428059333851418   
  
For EXP time: 54sec  
$T_1$ is: 426.36455000000007   
VOLT in $T_1$ is: 0.11353689999999998   
$T_2$ is: 482.9153000000001   
VOLT in $T_2$ is: 0.11709675000000001   
$T_m$ is: 457.98525000000006   
VOLT $T_m$ is: 0.23162735000000004   
$T_2-T_1$=56.55075000000005   
$T_2-T_m$=24.93005000000005   
$(T_2-T_m)/(T_2-T_1)$=0.44084384380401725   
  
For EXP time: 60sec  
$T_1$ is: 426.28799999999984   
VOLT in $T_1$ is: 0.09245025   
$T_2$ is: 482.43494999999996   
VOLT in $T_2$ is: 0.09609279999999998   
$T_m$ is: 458.12885000000006   
VOLT $T_m$ is: 0.18578665   
$T_2-T_1$=56.14695000000012   
$T_2-T_m$=24.3060999999999   
$(T_2-T_m)/(T_2-T_1)$=0.4329015200291352   
  
For EXP time: 90sec  
$T_1$ is: 426.96504999999996   
VOLT in $T_1$ is: 0.14553255000000004   
$T_2$ is: 483.04940000000005   
VOLT in $T_2$ is: 0.15246720000000005   
$T_m$ is: 457.3173500000001   
VOLT $T_m$ is: 0.29394975   
$T_2-T_1$=56.084350000000086   
$T_2-T_m$=25.73204999999996   
$(T_2-T_m)/(T_2-T_1)$=0.45880981057995535   

For EXP time: 150sec  
$T_1$ is: 426.6998   
VOLT in $T_1$ is: 0.22305419999999998   
$T_2$ is: 482.4951000000001   
VOLT in $T_2$ is: 0.23395379999999996   
$T_m$ is: 457.7724999999999   
VOLT $T_m$ is: 0.45437320000000003   
$T_2-T_1$=55.79530000000011   
$T_2-T_m$=24.72260000000017   
$(T_2-T_m)/(T_2-T_1)$=0.4430946692642592   
  
For EXP time: 210sec  
$T_1$ is: 426.47104999999993   
VOLT in $T_1$ is: 0.25906409999999996   
$T_2$ is: 483.17245   
VOLT in $T_2$ is: 0.26502004999999995   
$T_m$ is: 458.13244999999995   
VOLT $T_m$ is: 0.52695665   
$T_2-T_1$=56.70140000000009   
$T_2-T_m$=25.040000000000077   
$(T_2-T_m)/(T_2-T_1)$=0.44161167096403325   
  
For EXP time: 240sec  
$T_1$ is: 428.42440000000005   
VOLT in $T_1$ is: 0.2601295   
$T_2$ is: 482.9138499999999   
VOLT in $T_2$ is: 0.27255720000000005   
$T_m$ is: 459.15795000000014   
VOLT $T_m$ is: 0.53034195   
$T_2-T_1$=54.48944999999986   
$T_2-T_m$=23.75589999999977   
$(T_2-T_m)/(T_2-T_1)$=0.43597246806491585   
  
For EXP time: 300sec  
$T_1$ is: 426.9743500000001   
VOLT in $T_1$ is: 0.28164540000000005   
$T_2$ is: 482.15055   
VOLT in $T_2$ is: 0.28307965   
$T_m$ is: 458.03950000000003   
VOLT $T_m$ is: 0.5637149499999999   
$T_2-T_1$=55.17619999999994   
$T_2-T_m$=24.111049999999977   
$(T_2-T_m)/(T_2-T_1)$=0.43698279330580947   
  
For EXP time: 360sec  
$T_1$ is: 427.71555000000006   
VOLT in $T_1$ is: 0.2932864   
$T_2$ is: 482.84815   
VOLT in $T_2$ is: 0.3019807   
$T_m$ is: 458.6387000000001   
VOLT $T_m$ is: 0.58921055   
$T_2-T_1$=55.13259999999991   
$T_2-T_m$=24.20944999999989   
$(T_2-T_m)/(T_2-T_1)$=0.4391131562814003   
  
For EXP time: 420sec  
$T_1$ is: 426.41370000000006   
VOLT in $T_1$ is: 0.2923916   
$T_2$ is: 482.2306   
VOLT in $T_2$ is: 0.30460989999999993   
$T_m$ is: 457.8101   
VOLT $T_m$ is: 0.5968545999999999   
$T_2-T_1$=55.81689999999992   
$T_2-T_m$=24.420500000000004   
$(T_2-T_m)/(T_2-T_1)$=0.43751086140577566   
  
For EXP time: 480sec  
$T_1$ is: 426.50165000000004   
VOLT in $T_1$ is: 0.27646930000000003   
$T_2$ is: 482.59010000000006   
VOLT in $T_2$ is: 0.28325455   
$T_m$ is: 457.0970000000001   
VOLT $T_m$ is: 0.5606833000000001   
$T_2-T_1$=56.08845000000002   
$T_2-T_m$=25.49309999999997   
$(T_2-T_m)/(T_2-T_1)$=0.45451603672413765   